packages/graphql_client/lib/src/farms.dart

```diff
@@ -1,46 +1,82 @@
+
     final body = '''
-query getFarms(\$twinId: Int!) {
-  farms(where: {twinID_eq: \$twinId}) {
-    farmID
-    name
-    certification
-    dedicatedFarm
-    pricingPolicyID
-    stellarAddress
-    twinID
+            query Farms{
+              farms $queryString {
+                $returnString
+              }
+            }''';
+
+    final response = await gqlClient.query(body);
+    if (response['data'] == null || response['data']['farms'] == null) {
+      throw Exception('Invalid response structure: $response');
+    }
+    final farmsDataList = response['data']['farms'] as List<dynamic>;
+    if (farmsDataList.any((item) => item is! Map<String, dynamic>)) {
+      throw Exception(
+          'Invalid data format: Expected a list of maps: $response');
+    }
+    List<FarmInfo> farms =
+        (response['data']['farms'] as List<dynamic>).map((farmsData) {
+      return FarmInfo(
+        id: farmsData['id'],
+        farmID: farmsData['farmID'] ?? 0,
+        certification: farmsData['certification'] ?? '',
+        dedicatedFarm: farmsData['dedicatedFarm'] ?? false,
+        name: farmsData['name'] ?? '',
+        pricingPolicyID: farmsData['pricingPolicyID'] ?? 0,
+        stellarAddress: farmsData['stellarAddress'] ?? '',
+        twinID: farmsData['twinID'] ?? 0,
+        publicIPs: farmsData['publicIPs'] != null
+            ? (farmsData['publicIPs'] as List<dynamic>)
+                .map((item) => PublicIpsInfo.fromJson(item))
+                .toList()
+            : null,
+      );
+    }).toList();
+    return farms;
   }
-}''';
-
-    try {
-      final response =
-          await gqlClient.query(body, variables: {'twinId': options.twinId});
-      List<FarmInfo> farms =
-          (response['data']['farms'] as List<dynamic>).map((farmsData) {
-        return FarmInfo(
-          farmID: farmsData['farmID'] ?? 0,
-          certification: farmsData['certification'] ?? '',
-          dedicatedFarm: farmsData['dedicatedFarm'] ?? false,
-          name: farmsData['name'] ?? '',
-          pricingPolicyID: farmsData['pricingPolicyID'] ?? 0,
-          stellarAddress: farmsData['stellarAddress'] ?? '',
-          twinID: farmsData['twinID'] ?? 0,
-        );
-      }).toList();
-
-      return farms;
-    } catch (err) {
-      throw err;
+
+  //sorts by id_ASC by default
+  //return total count by default
+  Future<FarmsConnectionInfo> getFarmsConnection(
+      FarmsConnectionQueryOptions? queryOptions,
+      FarmsConnectionReturnOptions? returnOptions) async {
+    final queryString = queryOptions?.toString() ?? "(orderBy: id_ASC)";
+    final returnString = returnOptions?.toString() ?? "totalCount \n";
+
+    final body = '''
+            query FarmsConnection{
+              farmsConnection $queryString {
+                $returnString
+              }
+            }''';
+
+    final response = await gqlClient.query(body);
+
+    if (response['data'] == null ||
```

Comment:
let's be consistent with all errors
```suggestion
     if (response['data'] == null) {
    throw Exception('Missing "data" field in response: $response');
  }
  if (response['data']['farmsConnection'] == null) {
    throw Exception('Missing "farmsConnection" field in response data: ${response['data']}');
  }

  final farmsConnectionData = response['data']['farmsConnection'];

  if (farmsConnectionData is! Map<String, dynamic>) {
    throw Exception('Invalid data format: Expected a Map<String, dynamic> for "farmsConnection" field, got ${farmsConnectionData.runtimeType}');
  }
```